pkg/command/context.go

```diff
@@ -1239,3 +1274,126 @@ func prepareClusterServerURL(context *configtypes.Context, projectName, spaceNam
+		return nil, cobra.ShellCompDirectiveNoFileComp
+	}
+
+	cfg, err := config.GetClientConfig()
+	if err != nil {
+		return nil, cobra.ShellCompDirectiveNoFileComp
+	}
+
+	target := getTarget()
+
+	var allCtxs []*configtypes.Context
+	for _, ctx := range cfg.KnownContexts {
+		if target == configtypes.TargetUnknown || target == ctx.Target {
+			allCtxs = append(allCtxs, ctx)
+		}
+	}
+	return completionFormatCtxs(allCtxs), cobra.ShellCompDirectiveNoFileComp
+}
+
+func completeTAEContexts(_ *cobra.Command, args []string, _ string) ([]string, cobra.ShellCompDirective) {
+	if len(args) > 0 {
+		return nil, cobra.ShellCompDirectiveNoFileComp
+	}
+
+	cfg, err := config.GetClientConfig()
+	if err != nil {
+		return nil, cobra.ShellCompDirectiveNoFileComp
+	}
+
+	var taeCtxs []*configtypes.Context
+	for _, ctx := range cfg.KnownContexts {
+		if ctx.Target == configtypes.TargetTAE {
+			taeCtxs = append(taeCtxs, ctx)
+		}
+	}
+	return completionFormatCtxs(taeCtxs), cobra.ShellCompDirectiveNoFileComp
+}
+
+func completeActiveContexts(_ *cobra.Command, args []string, _ string) ([]string, cobra.ShellCompDirective) {
+	if len(args) > 0 {
+		return nil, cobra.ShellCompDirectiveNoFileComp
+	}
+
+	currentCtxMap, err := config.GetAllCurrentContextsMap()
+	if err != nil {
+		return nil, cobra.ShellCompDirectiveNoFileComp
+	}
+
+	target := getTarget()
+
+	var allCtxs []*configtypes.Context
+	for _, ctx := range currentCtxMap {
+		if target == configtypes.TargetUnknown || target == ctx.Target {
+			allCtxs = append(allCtxs, ctx)
+		}
+	}
+	return completionFormatCtxs(allCtxs), cobra.ShellCompDirectiveNoFileComp
+}
+
+// Setup shell completion for the kube-context flag
+func completeKubeContext(_ *cobra.Command, _ []string, _ string) ([]string, cobra.ShellCompDirective) {
+	if kubeConfig == "" {
+		kubeConfig = getDefaultKubeconfigPath()
+	}
+
+	cobra.CompDebugln("About to get the different kube-contexts", false)
+
+	kubeclient, err := clientcmd.NewNonInteractiveDeferredLoadingClientConfig(
+		&clientcmd.ClientConfigLoadingRules{ExplicitPath: kubeConfig},
+		&clientcmd.ConfigOverrides{}).RawConfig()
+	if err != nil {
+		return nil, cobra.ShellCompDirectiveNoFileComp
+	}
+
+	var comps []string
+	for name, context := range kubeclient.Contexts {
+		comps = append(comps, fmt.Sprintf("%s\t%s@%s", name, context.AuthInfo, context.Cluster))
+	}
+	// Sort the completion to make testing easier
+	sort.Strings(comps)
+	return comps, cobra.ShellCompDirectiveNoFileComp
+}
+
+func completeCreateCtx(_ *cobra.Command, args []string, _ string) ([]string, cobra.ShellCompDirective) {
+	if len(args) == 0 {
+		comps := cobra.AppendActiveHelp(nil, "Please specify a name for the context")
+		return comps, cobra.ShellCompDirectiveNoFileComp
+	}
+
+	if endpoint == "" && kubeContext == "" {
+		// The user must provide more info by using flags.
+		// Note that those flags are not marked as mandatory
+		// because the prompt mechanism can be used instead.
+		comps := []string{"--"}
```

Comment:
more out of curiosity: what is the convent on when we should return "--"?
looks like we are doing this despite the flags not being mandatory.

How would this differ from flags accepted by, say, `tanzu context list`, which are also not mandatory?

Reply:
Good question.  It turns out that the flags _are_ mandator in the sense that if they are not specified by the user, the prompt will be triggered to request the flag information.  We cannot mark them as "required" in Cobra, or else the command would fail before the prompt was triggered.

For shell completion I decided that we should help the user build a complete command that would not trigger the prompt, which is why I added the `--` to get the flags values needed.  But I now realize it won't be clear _which_ flags are required...

Not sure what is best anymore.  Any thoughts?

Reply:
If we want to complete the flags, we could complete the specific flags that are required.  The complexity here is that it depends on what type of context the user wants to create to know which flags are required.